src/charm.py

```diff
@@ -759,19 +870,67 @@ def _reconcile_mongo_hosts_and_users(self, event: RelationEvent) -> None:
                 logger.info("Deferring reconfigure: error=%r", e)
                 event.defer()
 
+    def __handle_partition_on_stop(self) -> None:
+        """Raise partition to prevent other units from restarting if an upgrade is in progress.
+
+        If an upgrade is not in progress, the leader unit will reset the partition to 0.
+        """
+        current_unit_number = unit_number(self.unit)
+        if k8s_upgrade.partition.get(app_name=self.app.name) < current_unit_number:
+            k8s_upgrade.partition.set(app_name=self.app.name, value=current_unit_number)
+            logger.debug(f"Partition set to {current_unit_number} during stop event")
+
+    def __handle_relation_departed_on_stop(self) -> None:
+        """Leaves replicaset.
+
+        If the unit has not already left the replica set, this function
+        attempts to block operations until the unit is removed. Note that with
+        how Juju currently operates, we only have 30 seconds until SIGTERM
+        command, so we are by no means guaranteed to have removed the replica
+        before the pod is removed. However the leader will reconfigure the
+        replica set if this is the case on `update status`.
+        """
+        logger.debug(f"{self.unit.name} blocking on_stop")
+        is_in_replica_set = True
+        timeout = UNIT_REMOVAL_TIMEOUT
+        while is_in_replica_set and timeout > 0:
+            is_in_replica_set = self.is_unit_in_replica_set()
+            time.sleep(1)
+            timeout -= 1
+            if timeout < 0:
+                raise Exception(f"{self.unit.name}.on_stop timeout exceeded")
+        logger.debug("{self.unit.name} releasing on_stop")
+        self.unit_departed = False
+
+    def __handle_upgrade_on_stop(self) -> None:
+        """Sets the unit state to RESTARTING and step down from replicaset.
+
+        Note that with how Juju currently operates, we only have at most 30
+        seconds until SIGTERM command, so we are by no means guaranteed to have
+        stepped down before the pod is removed.
+        Upon restart, the upgrade will still resume because all hooks run the
+        `_reconcile_upgrade` handler.
+        """
+        self.upgrade._upgrade.unit_state = UnitState.RESTARTING
+
+        # According to the MongoDB documentation, before upgrading the primary, we must ensure a
+        # safe primary re-election.
+        try:
+            if self.unit.name == self.primary:
+                logger.debug("Stepping down current primary, before upgrading service...")
+                self.upgrade.step_down_primary_and_wait_reelection()
```

Comment:
@MiaAltieri `step_down_primary_and_wait_reelection` could use the statefulset patching

Reply:
You can't do that there:
 * At this moment, the pod has already received the shutdown notice, so patching it would change nothing.

If the statefulset is to be patched, I think it should be done manually by the juju admin, and it means it should be done everytime the statefulset restarts/rolls out. Also juju offers no guarantees that it will not patch again the value later.
I don't know what is the correct solution here, the best would be that the shutdown period is exposed in the model-config but it seems ([see the ticket](https://bugs.launchpad.net/juju/+bug/2035102)).